# Quiz: SQL Query

Selamat Anda sudah menyelesaikan materi *SQL Query*! Kami akan melakukan penilaian berupa kuis untuk menguji materi yang sudah dipelajari.

## Skema Database

Pada kuis ini, database yang digunakan adalah `chinook.db` yang terdapat pada folder `data_input`.

Database `chinook.db` terdiri dari 11 tabel yang berisi sampel data dari toko media digital:

- Data terkait media seperti pada tabel `tracks`, `albums`, `artists`, `genres`, `media_types`, `playlists`, `playlist_track` merupakan data asli dari *library Apple iTunes*.
- Informasi mengenai `customers` dan `employees` dibuat menggunakan nama dan alamat fiktif yang dapat ditemukan di *Google Maps*, dan data lain yang diformat dengan baik (telepon, fax, email, dll).
- Informasi penjualan `invoices` dan `invoice_items` dihasilkan secara otomatis menggunakan data acak untuk periode tahun 2009-2013.

In [1]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("data_input/chinook.db")

![](assets/chinookschema2.png)

### SQL Query

Pertama-tama mari kita coba untuk membuat dataframe menggunakan method `read_sql_query()`, anda ingin mengetahui total pembelian yang dilakukan oleh masing-masing pelanggan. Untuk mengetahui nama pelanggan, Anda perlu menggabungkan nama depan (`FirstName`) dan nama belakang (`LastName`) pelanggan dan simpanlah kedalam kolom baru dengan nama `CustomerName`. Buatlah tabel agregasi yang menunjukkan **jumlah** total penjualan (`Total`) pada masing-masing pelanggan dan urutkanlah untuk mendapatkan informasi total pembelian terbesar hingga terkecil.

```python
query = '''
    SELECT SUM(invoices.total) as Total, customers.firstname || ' ' || customers.___lastname AS CustomerName  
    FROM customers

    LEFT JOIN invoices
    ON customers.customerid = invoices.customerid

    GROUP BY CustomerName
    ORDER BY Total DESC

    LIMIT 5 
'''
    
pd.read_sql_query(query, conn)
```


1. Manakah di antara pelanggan berikut yang termasuk dalam 5 besar pelanggan yang diurutkan berdasarkan total pembeliannya (`Total`)?
    - [ ] Victor Stevens
    - [ ] Hugh O'Reilly
    - [ ] Puja Srivastava
    - [ ] Astrid Gruber

In [3]:
## Your code here
query = '''
    SELECT SUM(invoices.total) as Total, customers.firstname || ' ' || customers.lastname AS CustomerName  
    FROM customers

    LEFT JOIN invoices
    ON customers.customerid = invoices.customerid

    GROUP BY CustomerName
    ORDER BY Total DESC

    LIMIT 5 
'''
    
pd.read_sql_query(query, conn)


,Total,CustomerName
0,49.62,Helena Holý
1,47.62,Richard Cunningham
2,46.62,Luis Rojas
3,45.62,Ladislav Kovács
4,45.62,Hugh O'Reilly


Anda ingin mengetahui pendapatan bruto yang diperoleh oleh masing-masing negara. Lakukanlah fecthing data dari database untuk mendapatkan **jumlah** dari pendapatan bruto (`Total`) yang diperoleh oleh masing-masing negara dan urutkanlah agar mendapatkan informasi negara yang memiliki pendapatan paling besar hingga paling kecil.

```python
query = '''
    SELECT SUM(invoices.total) as Total, customers.country  
    FROM customers

    LEFT JOIN invoices
    ON customers.customerid = invoices.customerid

    GROUP BY country
    ORDER BY Total DESC
'''
    
pd.read_sql_query(query, conn)
```

2. Manakah di antara negara berikut yang termasuk dalam 5 negara dengan pendapatan bruto paling besar yang diurutkan berdasarkan kolom `Total`?
    - [ ] United Kingdom
    - [ ] Germany
    - [ ] Portugal
    - [ ] India

In [5]:
## Your code here
query = '''
    SELECT SUM(invoices.total) as Total, customers.country  
    FROM customers

    LEFT JOIN invoices
    ON customers.customerid = invoices.customerid

    GROUP BY country
    ORDER BY Total DESC
    LIMIT 5
'''
    
pd.read_sql_query(query, conn)



,Total,Country
0,523.06,USA
1,303.96,Canada
2,195.10,France
3,190.10,Brazil
4,156.48,Germany


Berdasarkan informasi yang sudah Anda dapatkan mengenai 5 negara dengan pendapatan tertinggi, Anda ingin melakukan analisa lebih lanjut beberapa informasi lain yang bisa dijadikan sebagai *insight*. Gunakan kerangka kode di bawah ini sebagai referensi untuk membuat sebuah dataframe yang berisi catatan pelanggan-pelanggan yang berasal dari 5 negara dengan total pendapatan teratas. 

```python
query = '''
    SELECT ___.Total, ___.InvoiceDate, customers.country, 
    customers.____ || ' ' || customers.____ as CustomerName
    FROM customers
    LEFT JOIN ____
    ON customers.____ = ____.____
    WHERE ____.____ IN ('___' , '___', '___', '___','___')
'''

top5data = pd.read_sql_query(query, conn, parse_dates='InvoiceDate')
```

> **Petunjuk** : Pada method `read_sql_query`, Anda dapat menggunakan argumen `parse_dates='InvoiceDate'` agar kolom tertentu dikonversikan sebagai tanggal, sehingga Anda tidak perlu lagi melakukan konversi `to_datetime ()`

In [7]:
## Your code here
query = '''
    SELECT invoices.total, invoices.invoicedate, customers.country, 
    customers.firstname || ' ' || customers.lastname as CustomerName
    FROM customers
    LEFT JOIN invoices
    ON customers.customerid = invoices.customerid
    WHERE customers.country IN ('USA' , 'Canada', 'France', 'Brazil','Germany')
'''

top5data = pd.read_sql_query(query, conn, parse_dates='InvoiceDate')
top5data

,Total,InvoiceDate,Country,CustomerName
0,3.98,2010-03-11,Brazil,Luís Gonçalves
1,3.96,2010-06-13,Brazil,Luís Gonçalves
2,5.94,2010-09-15,Brazil,Luís Gonçalves
3,0.99,2011-05-06,Brazil,Luís Gonçalves
4,1.98,2012-10-27,Brazil,Luís Gonçalves
...,...,...,...,...
240,5.94,2010-07-15,France,Isabelle Mercier
241,0.99,2011-03-05,France,Isabelle Mercier
242,1.98,2012-08-26,France,Isabelle Mercier
243,16.86,2012-10-06,France,Isabelle Mercier


Menggunakan dataframe `top5data` diatas, lakukan analisa deskriptif untuk menjawab pertanyaan dibawah ini.

3. Berapa banyak transaksi penjualan (baris) yang berasal dari 5 negara teratas?
    - [ ] 7
    - [ ] 245
    - [ ] 412
    - [ ] 1715

In [8]:
## Your code here

top5data

,Total,InvoiceDate,Country,CustomerName
0,3.98,2010-03-11,Brazil,Luís Gonçalves
1,3.96,2010-06-13,Brazil,Luís Gonçalves
2,5.94,2010-09-15,Brazil,Luís Gonçalves
3,0.99,2011-05-06,Brazil,Luís Gonçalves
4,1.98,2012-10-27,Brazil,Luís Gonçalves
...,...,...,...,...
240,5.94,2010-07-15,France,Isabelle Mercier
241,0.99,2011-03-05,France,Isabelle Mercier
242,1.98,2012-08-26,France,Isabelle Mercier
243,16.86,2012-10-06,France,Isabelle Mercier


Dengan menggunakan Pandas, kita dapat melakukan teknik-teknik data manipulasi seperti ekstraksi informasi dari tipe data datetime pada dataframe hasil fechting dari database. Contohnya, anda tertarik untuk mengetahui lebih lanjut penjualan yang dilakuan pada top 5 negara khususnya Perancis (**France**). Negara Perancis (**France**) melakukan penjualan setiap harinya, lakukan agregasi data untuk mendapatkan **jumlah** pendapatan masing-masing harinya.

Kita dapat menggunakan kolom `InvoiceDate` dari dataframe `top5data` untuk membuat kolom baru `InvoiceDOW` untuk menampilkan nama-nama hari pada suatu transaksi

> **Petunjuk** : `pandas` memiliki method bawaan untuk mengekstrak nama hari dalam seminggu. Kita telah mempelajarinya di materi ke-2 dari spesialisasi ini (**Working with Datetime**). Contoh penggunaannya adalah:
>
> `data['InvoiceDOW'] = data['InvoiceDate'].dt.day_name()`


In [ ]:
## Your code here



Selain Perancis, Anda juga tertarik untuk menganalisa pendapatan harian yang diperoleh dari top 5 negara dengan pendapatan tertinggi. Menggunakan dataframe `top5data`, lakukanlah agregasi untuk mengetahui **jumlah** penjualan masing-masing hari dari keseluruhan negara tersebut.


4. Berdasarkan agregasi data yang Anda lakukan diatas, jawablah 2 pertanyaan berikut: Berapa jumlah penjualan (`Total`) yang kita hasilkan dari total(`sum`) pasar Perancis (**France**) pada hari Selasa? Berapa total penjualan yang kita hasilkan dari semua gabungan 5 pasar teratas pada hari Jumat?
    - [ ] 160.41 / 173.31  
    - [ ] 160.41 / 326.77
    - [ ] 1.98 / 173.31
    - [ ] 1.98 / 326.77

In [18]:
## Your code here
query = '''
    SELECT
      case cast (strftime('%w', invoicedate) as integer)
    when 0 then 'Sunday'
    when 1 then 'Monday'
    when 2 then 'Tuesday'
    when 3 then 'Wednesday'
    when 4 then 'Thursday'
    when 5 then 'Friday'
    END AS DayOfWeek,
        SUM(total) as TotalSales
    FROM (
    SELECT invoices.total, invoices.invoicedate, customers.country, 
    customers.firstname || ' ' || customers.lastname as CustomerName
    FROM customers
    LEFT JOIN invoices
    ON customers.customerid = invoices.customerid
    WHERE customers.country IN ('USA' , 'Canada', 'France', 'Brazil','Germany'))
    WHERE country = 'France' 
    and DayOfWeek = 'Tuesday'
    GROUP BY DayOfWeek
    ORDER BY TotalSales DESC
'''

top5data = pd.read_sql_query(query, conn, parse_dates='InvoiceDate')
top5data

,DayOfWeek,TotalSales
0,Tuesday,1.98


In [28]:
## Your code here
query = '''
    SELECT
      case cast (strftime('%w', invoicedate) as integer)
    when 0 then 'Sunday'
    when 1 then 'Monday'
    when 2 then 'Tuesday'
    when 3 then 'Wednesday'
    when 4 then 'Thursday'
    when 5 then 'Friday'
    END AS DayOfWeek,
        SUM(total) as TotalSales
    FROM (
    SELECT invoices.total, invoices.invoicedate, customers.country, 
    customers.firstname || ' ' || customers.lastname as CustomerName
    FROM customers
    LEFT JOIN invoices
    ON customers.customerid = invoices.customerid
    WHERE customers.country IN ('USA' , 'Canada', 'France', 'Brazil','Germany'))
    WHERE DayOfWeek = 'Friday'
    GROUP BY DayOfWeek
    ORDER BY TotalSales DESC
'''

top5data = pd.read_sql_query(query, conn, parse_dates='InvoiceDate')
top5data

,DayOfWeek,TotalSales
0,Friday,173.31


Perusahaan ingin mengetahui daftar pelanggan yang pernah membeli produk dengan harga lebih tinggi daripada rata-rata harga semua produk. Tim analisis mempertimbangkan untuk menggunakan subquery.

5. Mengapa subquery cocok digunakan pada kasus tersebut?

    - [ ] Karena subquery dapat mengganti semua kebutuhan JOIN.
    - [ ] Karena subquery bisa menghitung rata-rata produk terlebih dahulu, lalu hasilnya dipakai untuk filter data pelanggan.
    - [ ] Karena subquery selalu menghasilkan tabel permanen yang bisa dipakai ulang.
    - [ ] Karena subquery secara otomatis mengurutkan data sesuai kebutuhan analisis.

Perusahaan ingin mengurutkan setiap pelanggan berdasarkan frekuensi pembelian dalam kelompok kota masing-masing menggunakan window function.

6. Mengapa window function lebih unggul dibanding GROUP BY dalam kasus tersebut?

    - [ ] GROUP BY menghilangkan detail tiap pelanggan, sedangkan window function mempertahankan semua baris dengan ranking tambahan.
    - [ ] Window function otomatis menambahkan ranking dalam tabel.
    - [ ] GROUP BY tidak bisa digunakan untuk agregasi per kota, sedangkan window function dapat mengagregasi per kota.
    - [ ] Window function lebih cepat dibandingkan GROUP BY.